src/components/time-picker/time-picker.tsx

```diff
@@ -0,0 +1,374 @@
+                        />
+                    </label>
+                }
+            </div>
+        );
+    }
+
+    private getValue(): string {
+        const {hh = 0, mm = 0, ampm} = this.state;
+        const hhString = formatTimeChunk(to24(Number(hh), ampm));
+        const mmString = formatTimeChunk(mm);
+        return `${hhString}:${mmString}`;
+    }
+
+    private select(segment: Segment) {
+        const input = this.segments[segment]!;
+        this.setState({
+            currentSegment: segment
+        }, () => {
+            input.focus();
+            if (isTimeSegment(segment)) {
+                (input as HTMLInputElement).select();
+            }
+        });
+    }
+    private moveSelection(step: number): boolean {
+        const index = segments.indexOf(this.state.currentSegment!);
+        const nextSegment = segments[index + step];
+        if (this.segments[nextSegment]) {
+            this.select(nextSegment);
+            return true;
+        }
+        return false;
+    }
+
+    private updateSegmentValue(name: Segment, value: string | Ampm): void {
+        this.setState({
+            [name as any]: value
+        }, () => {
+            if (isTimeSegment(name)) {
+                this.select(name);
+            }
+            this.commit();
+        });
+    }
+
+    private toggleAmpm(shouldSelect: boolean): void {
+        this.updateSegmentValue('ampm', ampmSwitch[this.state.ampm]);
+        if (shouldSelect) {
+            this.segments.ampm!.focus();
+        } else {
+            this.segments.ampm!.blur();
+        }
+    }
+
+    private changeValue(step: number): void {
+        const {currentSegment, ampm} = this.state;
+        if (!isTimeSegment(currentSegment)) {
+            return this.toggleAmpm(true);
+        }
+        let newValue: any = Number(this.state[currentSegment] || 0) + step;
+        newValue = getCircularValue(currentSegment, newValue, ampm);
+        newValue = formatTimeChunk(newValue);
+        this.updateSegmentValue(currentSegment, newValue);
+    }
+
+    private onPlaceholderClick = () => this.segments.hh!.focus();
+
+    private onStepperUp = () => this.changeValue(1);
+    private onStepperDown = () => this.changeValue(-1);
+
+    private onAmpmMouseDown = (e: React.SyntheticEvent<HTMLDivElement>) => {
+        e.preventDefault();
+        if (!this.props.disabled) {
+            this.toggleAmpm(document.activeElement === this.segments.ampm);
+        }
+    }
+
+    private onAmpmFocus = (e: React.SyntheticEvent<HTMLDivElement>) => {
+        this.setState({focus: true, currentSegment: 'ampm'});
+    }
+
+    private onInputMouseDown = (e: any): void => {
+        e.preventDefault();
+        e.currentTarget.focus();
+    }
+
+    private onInputChange = (e: React.SyntheticEvent<HTMLInputElement>): void => {
+        const {ampm} = this.state;
+        const {value} = e.currentTarget;
+        const name = e.currentTarget.name as TimeSegment;
```

Comment:
type guard could help here only if I will wrap all code in `if (isTimeSegment(name))` but I don't want to this, because `name` here is always a `TimeSegment`